### PR TITLE
Set cuda device before create cuda stream for IOBinding case

### PR DESCRIPTION
**Pull request by jslhcl**
### Description

Set cuda device before create cuda stream for IOBinding case

### Motivation and Context

This is to fix the issue #18432  , which the inference will fail for IOBinding case when there are multiple cuda devices. The reason is that the cuda device is not set properly before the cuda stream is created

### CODE REVIEW

File: onnxruntime/test/python/onnxruntime_test_python.py [hunk 1]
```diff
@@ -60,6 +60,35 @@ def run_model_with_input(self, session_object, input_name, input_value, iter_num
             predict = session_object.run(None, {input_name: input_value})[0]
             queue.put(max(predict.flatten().tolist()))
 
+    def load_cuda_lib(self):
```
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/test/python/onnxruntime_test_python.py:63]: yes, tested the two test cases which invoke this function

[the remark this answers is not in the document]
Comment (snnn) [reply onnxruntime/test/python/onnxruntime_test_python.py:63]: On MacOS?

[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/test/python/onnxruntime_test_python.py:63]: On Linux

[the remark this answers is not in the document]
Comment (snnn) [reply onnxruntime/test/python/onnxruntime_test_python.py:63]: But this code has three branches.  Should you test them all?

[the remark this answers is not in the document]
Comment (snnn) [reply onnxruntime/test/python/onnxruntime_test_python.py:63]: No, I do not insist you doing this.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 2]
```diff
@@ -383,8 +383,6 @@ Status CUDAExecutionProvider::Sync() const {
 }
 
 Status CUDAExecutionProvider::OnRunStart() {
-  // always set CUDA device when session::Run() in case it runs in a worker thread
-  CUDA_RETURN_IF_ERROR(cudaSetDevice(GetDeviceId()));
```
Comment (souptc) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:387]: should we keep it? there may be other cuda component other than stream depends on current device id, like cudnn

Comment (jslhcl) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:387]: This change is essentially moving cudaSetDevice() from OnRunStart() to 

DeviceStreamCollectionHolder device_stream_collection_holder(session_state_.get());

there are only a few lines of codes between these 2 places (in InferenceSession::Run()) and no cuda related operation involved. So I think deletion here is safe.

**Review by souptc — APPROVED** [review]

**Pull request merged**